--- /dev/null	2026-07-30 18:45:23.488142685 +0000
+++ arch/arm64/boot/dts/qcom/msm8916-generic-uf02.dts	2026-07-30 20:43:07.682560659 +0000
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/dts-v1/;
+
+#include "msm8916-ufi.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+        model = "UF02 4G Modem Stick (HMU05 Test)";
+        compatible = "uf02,250605v0s", "qcom,msm8916";
+};
+
+&button_restart {
+        gpios = <&tlmm 37 GPIO_ACTIVE_LOW>;
+};
+
+&led_r {
+        gpios = <&tlmm 72 GPIO_ACTIVE_HIGH>;
+};
+
+&led_g {
+        function = LED_FUNCTION_WAN;
+        gpios = <&tlmm 71 GPIO_ACTIVE_HIGH>;
+};
+
+&led_b {
+        function = LED_FUNCTION_WLAN;
+        gpios = <&tlmm 73 GPIO_ACTIVE_HIGH>;
+};
+
+&button_default {
+        pins = "gpio37";
+        bias-pull-up;
+};
+
+&gpio_leds_default {
+        pins = "gpio71", "gpio72", "gpio73";
+};
+
+&tlmm {
+
+        sim1_en {
+                gpio-hog;
+                gpios = <119 GPIO_ACTIVE_HIGH>;
+                output-high;
+                line-name = "sim1_en";
+        };
+
+        sim2_en {
+                gpio-hog;
+                gpios = <14 GPIO_ACTIVE_HIGH>;
+                output-low;
+                line-name = "sim2_en";
+        };
+
+        sim3_en {
+                gpio-hog;
+                gpios = <12 GPIO_ACTIVE_HIGH>;
+                output-low;
+                line-name = "sim3_en";
+        };
+};
