--- a/drivers/net/wwan/qcom_bam_dmux.c
+++ b/drivers/net/wwan/qcom_bam_dmux.c
@@ -279,9 +279,15 @@ static int bam_dmux_netdev_open(struct net_device *netdev)
 	struct bam_dmux_netdev *bndev = netdev_priv(netdev);
 	int ret;
 
+	ret = pm_runtime_resume_and_get(bndev->dmux->dev);
+	if (ret < 0)
+		return ret;
+
 	ret = bam_dmux_send_cmd(bndev, BAM_DMUX_CMD_OPEN);
-	if (ret)
+	if (ret) {
+		pm_runtime_put(bndev->dmux->dev);
 		return ret;
+	}
 
 	netif_start_queue(netdev);
 	return 0;
@@ -293,6 +299,7 @@ static int bam_dmux_netdev_stop(struct net_device *netdev)
 
 	netif_stop_queue(netdev);
 	bam_dmux_send_cmd(bndev, BAM_DMUX_CMD_CLOSE);
+	pm_runtime_put(bndev->dmux->dev);
 	return 0;
 }
 
@@ -359,6 +366,8 @@ static netdev_tx_t bam_dmux_netdev_start_xmit(struct sk_buff *skb,
 		if (!atomic_long_fetch_or(BIT(skb_dma - dmux->tx_skbs),
 					  &dmux->tx_deferred_skb))
 			queue_pm_work(&dmux->tx_wakeup_work);
+		DEV_STATS_INC(netdev, tx_packets);
+		DEV_STATS_ADD(netdev, tx_bytes, skb->len);
 		return NETDEV_TX_OK;
 	}
 
@@ -366,6 +375,8 @@ static netdev_tx_t bam_dmux_netdev_start_xmit(struct sk_buff *skb,
 		goto drop;
 
 	dma_async_issue_pending(dmux->tx);
+	DEV_STATS_INC(netdev, tx_packets);
+	DEV_STATS_ADD(netdev, tx_bytes, skb->len);
 	return NETDEV_TX_OK;
 
 drop:
@@ -536,6 +547,9 @@ static void bam_dmux_cmd_data(struct bam_dmux_skb_dma *skb_dma)
 		break;
 	}
 
+	DEV_STATS_INC(netdev, rx_packets);
+	DEV_STATS_ADD(netdev, rx_bytes, skb->len);
+
 	netif_receive_skb(skb);
 }
 
@@ -671,6 +685,10 @@ static void bam_dmux_power_off(struct bam_dmux *dmux)
 	}
 
 	bam_dmux_free_skbs(dmux->rx_skbs, DMA_FROM_DEVICE);
+	bam_dmux_free_skbs(dmux->tx_skbs, DMA_TO_DEVICE);
+	dmux->tx_next_skb = 0;
+	atomic_long_set(&dmux->tx_deferred_skb, 0);
+	bitmap_zero(dmux->remote_channels, BAM_DMUX_NUM_CH);
 }
 
 static irqreturn_t bam_dmux_pc_irq(int irq, void *data)
