--- a/arch/arm64/boot/dts/qcom/msm8916.dtsi
+++ b/arch/arm64/boot/dts/qcom/msm8916.dtsi
@@ -144,8 +144,8 @@
 			interconnects = <&bimc MASTER_AMPSS_M0 &bimc SLAVE_EBI_CH0>;
 			interconnect-names = "cpu-mem";
 			#cooling-cells = <2>;
-			power-domains = <&CPU_PD0>;
-			power-domain-names = "psci";
+			power-domains = <&CPU_PD0>, <&cpr>;
+			power-domain-names = "psci", "cpr";
 			qcom,acc = <&cpu0_acc>;
 			qcom,saw = <&cpu0_saw>;
 		};
@@ -161,8 +161,8 @@
 			interconnects = <&bimc MASTER_AMPSS_M0 &bimc SLAVE_EBI_CH0>;
 			interconnect-names = "cpu-mem";
 			#cooling-cells = <2>;
-			power-domains = <&CPU_PD1>;
-			power-domain-names = "psci";
+			power-domains = <&CPU_PD1>, <&cpr>;
+			power-domain-names = "psci", "cpr";
 			qcom,acc = <&cpu1_acc>;
 			qcom,saw = <&cpu1_saw>;
 		};
@@ -178,8 +178,8 @@
 			interconnects = <&bimc MASTER_AMPSS_M0 &bimc SLAVE_EBI_CH0>;
 			interconnect-names = "cpu-mem";
 			#cooling-cells = <2>;
-			power-domains = <&CPU_PD2>;
-			power-domain-names = "psci";
+			power-domains = <&CPU_PD2>, <&cpr>;
+			power-domain-names = "psci", "cpr";
 			qcom,acc = <&cpu2_acc>;
 			qcom,saw = <&cpu2_saw>;
 		};
@@ -195,8 +195,8 @@
 			interconnects = <&bimc MASTER_AMPSS_M0 &bimc SLAVE_EBI_CH0>;
 			interconnect-names = "cpu-mem";
 			#cooling-cells = <2>;
-			power-domains = <&CPU_PD3>;
-			power-domain-names = "psci";
+			power-domains = <&CPU_PD3>, <&cpr>;
+			power-domain-names = "psci", "cpr";
 			qcom,acc = <&cpu3_acc>;
 			qcom,saw = <&cpu3_saw>;
 		};
@@ -242,20 +242,23 @@
 	};
 
 	cpu_opp_table: opp-table-cpu {
-		compatible = "operating-points-v2";
+		compatible = "operating-points-v2-kryo-cpu";
 		opp-shared;
 
 		opp-200000000 {
 			opp-hz = /bits/ 64 <200000000>;
 			opp-peak-kBps = <3200000>;
+			required-opps = <&cpr_opp1>;
 		};
 		opp-400000000 {
 			opp-hz = /bits/ 64 <400000000>;
 			opp-peak-kBps = <3200000>;
+			required-opps = <&cpr_opp2>;
 		};
 		opp-800000000 {
 			opp-hz = /bits/ 64 <800000000>;
 			opp-peak-kBps = <8528000>;
+			required-opps = <&cpr_opp3>;
 		};
 	};
 
@@ -264,17 +267,17 @@
 
 		cpr_opp1: opp1 {
 			opp-level = <1>;
-			qcom,opp-fuse-level = <1>;
+			qcom,opp-fuse-level = <1>;	/* SVS */
 		};
 
 		cpr_opp2: opp2 {
 			opp-level = <2>;
-			qcom,opp-fuse-level = <2>;
+			qcom,opp-fuse-level = <1>;	/* SVS */
 		};
 
 		cpr_opp3: opp3 {
 			opp-level = <3>;
-			qcom,opp-fuse-level = <3>;
+			qcom,opp-fuse-level = <2>;	/* NORMAL */
 		};
 	};
 
